File: src/nylon/nylonobjects/NylonFunction.java
package nylon.nylonobjects;

import nylon.InlineFunction;
import nylon.NylonException;
import nylon.NylonObject;

import java.util.Stack;

/**
 * Created by Aedan Smith.
 */

@SuppressWarnings("unchecked")
public abstract class NylonFunction extends NylonObject<NylonFunction> {
    public NylonFunction(String id) {
        super(null, Type.FUNCTION, id);
        this.value = this;
    }

    public static String format(String s) {
        String v = "";
        int tDepth = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '[') {
                v += s.charAt(i);
                tDepth++;
                v += '\n';
                for (int k = 0; k < tDepth; k++) {
                    v += '\t';
                }
            } else if (s.charAt(i) == ']') {
                tDepth--;
                v += '\n';
                for (int k = 0; k < tDepth; k++) {
                    v += '\t';
                }
                v += s.charAt(i);
            } else if (s.charAt(i) == ',') {
                v += s.charAt(i);
                v += '\n';
                for (int k = 0; k < tDepth; k++) {
                    v += '\t';
                }
                i++;
            } else if (s.charAt(i) != '\n' && s.charAt(i) != '\t') {
                v += s.charAt(i);
            }
        }
        return v;
    }

    public void apply(Stack<NylonObject> stack) throws NylonException {
        try {
            applyImpl(stack);
        } catch (NylonException e) {
            e.add(this);
            throw e;
        } catch (Throwable e) {
            throw new NylonException(e.getClass().getCanonicalName() + ": " + e.getMessage(), this);
        }
    }

    protected abstract void applyImpl(Stack<NylonObject> stack) throws NylonException;

    public Stack<NylonObject> toStack() throws NylonException {
        Stack<NylonObject> stack = new Stack<>();
        this.apply(stack);
        return stack;
    }

    @Override
    public double toDouble() throws NylonException {
        return toStack().peek().toDouble();
    }

    @Override
    public long toLong() throws NylonException {
        return toStack().peek().toLong();
    }

    @Override
    public char toCharacter() throws NylonException {
        return toStack().peek().toCharacter();
    }

    @Override
    public boolean toBoolean() throws NylonException {
        return toStack().peek().toBoolean();
    }

    @Override
    public NylonArray toArray() throws NylonException {
        return new NylonArray(toStack());
    }

    @Override
    public NylonString toNylonString() throws NylonException {
        return toStack().peek().toNylonString();
    }

    @Override
    public NylonFunction toFunction() {
        return this;
    }

    @Override
    public NylonFunction concatenate(NylonObject object) {
        return new InlineFunction("ConcatenatedFunction(" + this.id + ", " + object.getId() + ")",
                this, object.toFunction());
    }

    @Override
    public NylonObject subtract(NylonObject object) throws NylonException {
        throw new NylonException("Cannot subtract functions", this);
    }

    @Override
    public NylonFunction clone() throws CloneNotSupportedException {
        return new NylonFunction(id) {
            @Override
            public void applyImpl(Stack<NylonObject> stack) throws NylonException {
                NylonFunction.this.applyImpl(stack);
            }

            @Override
            public String toString() {
                return NylonFunction.this.toString();
            }
        };
    }

    @Override
    public String getId() {
        return id;
    }

    @Override
    public String toString() {
        return this.getClass().getSimpleName();
    }
}

File: src/nylon/NylonFunction.kt
package nylon

import java.util.*
import java.util.function.Supplier

/**
 * Created by Aedan Smith.
 */

abstract class NylonFunction(val string: String, var argNum: Int = 0) {
    protected var args = ArrayList<NylonFunction>(argNum)

    protected abstract fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>)

    fun apply(stack: NylonStack, args: ArrayList<NylonFunction>) {
        try {
            applyImpl(stack, args)
        } catch (e: Exception) {
            throw NylonException(e, this)
        }
    }

    fun apply(stack: NylonStack) = apply(stack, args)

    fun apply(): NylonStack {
        val stack = NylonStack()
        apply(stack)
        return stack
    }

    fun resolveArgs(argumentSupplier: Supplier<NylonFunction>) {
        while (args.size < argNum) {
            args.add(argumentSupplier.get())
        }
    }

    fun resolveNestedArgs(argumentSupplier: Supplier<NylonFunction>) {
        args.forEach { it.resolveNestedArgs(argumentSupplier) }
        resolveArgs(argumentSupplier)
    }

    fun clone(): NylonFunction = object : NylonFunction(string, argNum) {
        init {
            this@NylonFunction.args.forEach { this.args.add(it.clone()) }
        }

        override fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>) {
            this@NylonFunction.applyImpl(stack, args)
        }

        override fun toString(args: ArrayList<NylonFunction>): String {
            return this@NylonFunction.toString(args)
        }
    }

    open fun toString(args: ArrayList<NylonFunction>): String {
        var s = string
        args.forEach { s += "$it" }
        return s
    }

    override fun toString() = toString(args)
}

fun createProvider(nylonObject: NylonObject<*>, string: String) = object : NylonFunction(string) {
    override fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>) {
        stack.push(nylonObject.clone())
    }
}

fun concatenate(vararg nylonFunctions: NylonFunction) = concatenate(nylonFunctions.map { it })

fun concatenate(nylonFunctions: List<NylonFunction>) = object : NylonFunction("") {
    init {
        args.addAll(nylonFunctions)
    }

    override fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>) {
        args.forEach { it.apply(stack) }
    }

    override fun toString(args: ArrayList<NylonFunction>): String {
        var s = string
        args.forEach { s += " "; s += it }
        return "{${if (s.isNotEmpty()) s.substring(1) else s}}"
    }
}

fun emptyFunction(): NylonFunction = object : NylonFunction("{}") {
    override fun applyImpl(stack: NylonStack, args: ArrayList<NylonFunction>) {}
}
